### PR TITLE
Lesson 3: Triangle Intro

### PR DESCRIPTION
**Pull request by Lokathor**
* [x] install [shaderc-rs](https://github.com/google/shaderc-rs), add to README and CI
* [x] DescriptorSetLayout
* [x] PipelineLayout
* [x] GraphicsPipeline
  * [x] Pass through vertex shader
  * [x] Fixed color fragment shader
  * Postpone as much shader stuff as possible until a later lesson.
* [x] Buffer basics (so we can upload them triangles)
* [x] Uploading Vertex Data
* [x] Mouse Controls The Triangle (via Push Constant?)
* [x] Discuss Immediate API vs Retained API
* [x] Discuss that the `shaderc-rs` binaries must be shipped separately if you want to use them at runtime.
* [x] Allow the window to be resized
* [x] Clamp the extent size so it's not crazy when using Metal
* [x] Write the whole stupid tutorial

### CODE REVIEW

**Comment by Lokathor** [comment]
https://github.com/grovesNL/gfx-hal-tutorial/blob/master/src/15_hello_triangle.rs#L382

**Comment by StarArawn** [comment]
### Step 1: Loading in the shader.
- [ ] Decide on shaderc or glsl-to-spirv
- [ ] Create a `GraphicsShaderSet` with two entry points(vert, frag).

### Step 2: Creating a pipeline.
Descriptor set layouts are not required for a triangle so creating a pool would look something like:

- [ ] Create empty pipeline layout.
- [ ] Create pipeline description using the `GraphicsShaderSet`. Note: This might be where you can swap over to wireframe by using: `gfx_hal::Primitive::LineList`. Ideally you would want to tell the frag shader that it should use the color white or something for the lines.
- [ ] Using the pipeline description add information about blend targets and your vertex layout.
- [ ] Use all of the above to create the actual pipeline.

### Step 3 Create the vertex buffer. This part is fairly easy.
- [ ] Create buffer
- [ ] Create memory
- [ ] Copy CPU vertex data to GPU using the mapping writer.

### Step 4: Create the `draw_triangles_frame` function.
I would suggest that this be done in a slightly different way from how you have the `draw_clear_frame` currently. Something more like:

#### Draw
Gets all the command buffer stuff ready and creates the encoder with: `begin_render_pass_inline` After the draw_triangles_frame gets called it should also submit the command buffers.

#### Draw triangles frame
This should actually render the triangle and receive the encoder from: `begin_render_pass_inline` 
- [ ] Binds the correct pipeline using: `encoder. bind_graphics_pipeline`.
- [ ] Binds the vertex buffer using: `encoder.bind_vertex_buffers`
- [ ] Call `encoder.draw`.

**Comment by Lokathor** [comment]
Yeah, on discord I said that we'd want a new method like

```rust
fn HalState::draw_traingles_frame(&mut self, triangles: &[Triangle]) -> Result<(), &'static str>;
```

But the more I thought about it the more I realized that such a method doesn't transition cleanly to more advanced forms. You _could_ use something like that, but you're likely to be having many connected triangles as a single model in terms of verticies and indexes.

So I think we'll target a single triangle:

```rust
fn HalState::draw_traingle_frame(&mut self, triangle: Triangle) -> Result<(), &'static str>;
```

With dynamic clear color (as before) and solid black for the triangle interior.

---

Also, I had asked about a "wireframe mode", but this is sufficiently more complex than OGL that I think we can skip on it.

**Comment by Lokathor** [comment]
I also agree that we want to make the setup _around_ writing the CommandBuffer as uniform as possible. Probably we'd write a single private method similar to a "map" style call, where you pass it a closure that accepts `&mut CommandBuffer`, it does all the setup, calls the closure, then does the submission. Then each `draw_whatever_frame` method can reuse that.

**Comment by Lokathor** [comment]
After a few false starts and help from @rukai we're on track to use `shaderc-rs` both in development and in the CI for all three major backends.

* Drawbacks: The shader compiler that `shaderc-rs` builds is _distinct_ from the binary that we build. If we want to build new shaders at runtime in a binary that we ship we'll have to package those compilers with our program as part of the distribution. I'm adding a mention of that to the TODOs for this lesson.

**Comment by StarArawn** [comment]
> After a few false starts and help from @rukai we're on track to use `shaderc-rs` both in development and in the CI for all three major backends.
> 
> * Drawbacks: The shader compiler that `shaderc-rs` builds is _distinct_ from the binary that we build. If we want to build new shaders at runtime in a binary that we ship we'll have to package those compilers with our program as part of the distribution. I'm adding a mention of that to the TODOs for this lesson.

I would also add something to the tutorial about how you can use shaderc to build your shaders and save them as spirv. Ideally your game engine would have some sort of asset package manager/system. which would handle compiling shaders, building mip-maps for textures, etc.

**Comment by Lokathor** [comment]
The code works now, but it should probably get cleanup before we write a tutorial for it

File: examples/triangle_intro.rs [hunk 1]
```diff
@@ -0,0 +1,847 @@
...
+  render_pass: ManuallyDrop<<back::Backend as Backend>::RenderPass>,
+  render_area: Rect,
+  queue_group: QueueGroup<back::Backend, Graphics>,
+  swapchain: ManuallyDrop<<back::Backend as Backend>::Swapchain>,
+  device: ManuallyDrop<back::Device>,
+  _adapter: Adapter<back::Backend>,
+  _surface: <back::Backend as Backend>::Surface,
+  _instance: ManuallyDrop<back::Instance>,
+}
+impl HalState {
```
[the remark this answers is not in the document]
Comment (Lokathor) [reply examples/triangle_intro.rs]: Yeah we can do those

File: examples/triangle_intro.rs [hunk 2]
```diff
@@ -0,0 +1,847 @@
...
+        .ok_or("Couldn't take ownership of the QueueGroup!")?;
+      if queue_group.queues.len() > 0 {
+        Ok(())
+      } else {
+        Err("The QueueGroup did not have any CommandQueues available!")
+      }?;
+      (device, queue_group)
+    };
+
+    // Create A Swapchain, this is extra long
```
[the remark this answers is not in the document]
Comment (Lokathor) [reply examples/triangle_intro.rs]: Mmmmmmaaaaaybe?

The vulkan-tutorial.com style breaks all sorts of things into their own little functions all over the place, and I certainly object to that style. As I say in the lesson side of things at one point, we're leaning toward [Style C](http://number-none.com/blow/john_carmack_on_inlined_code.html) code.

A piece of code being long isn't always enough to justify it being a stand alone function. I think that there'll be a lesson towards the end of 3D Basics where we try to break out the "reusable" parts and organize things before going into the more ad-hoc set of lessons, but until then it's what you might call "premature abstraction".

All that said, I'm a hypocrite and the `create_pipeline` series of operations is a separate function with no practical difference between that and the swapchain setup, other than perhaps "the pipeline setup is _even longer_".

**Comment by Lokathor** [comment]
We could block _further_ changes to the vscode and idea folders (I did in the latest commit), but the point of the current setting is to highlight the advice in the "clear the screen" lesson about how to setup RLS to properly work with the conditional dependencies that a gfx-hal program ends up having, which is otherwise fairly non-obvious.

**Pull request merged**